SDL/Event.cc

```diff
@@ -1813,20 +1813,42 @@ SDL::trackCandidatesBuffer<alpaka::DevCpu>* SDL::Event::getTrackCandidates()
+        alpaka::memcpy(queue, nTrackCanpT3Host_buf, trackCandidatesBuffers->nTrackCandidatespT3_buf, 1);
+        alpaka::memcpy(queue, nTrackCanpLSHost_buf, trackCandidatesBuffers->nTrackCandidatespLS_buf, 1);
+        alpaka::memcpy(queue, nTrackCanT5Host_buf, trackCandidatesBuffers->nTrackCandidatesT5_buf, 1);
         alpaka::wait(queue);
 
         int nTrackCanHost = *alpaka::getPtrNative(nTrackCanHost_buf);
-        trackCandidatesInCPU = new SDL::trackCandidatesBuffer<alpaka::DevCpu>(N_MAX_TRACK_CANDIDATES + N_MAX_PIXEL_TRACK_CANDIDATES, devHost, queue);
+        int nTrackCanpT5Host = *alpaka::getPtrNative(nTrackCanpT5Host_buf);
+        int nTrackCanpT3Host = *alpaka::getPtrNative(nTrackCanpT3Host_buf);
+        int nTrackCanpLSHost = *alpaka::getPtrNative(nTrackCanpLSHost_buf);
+        int nTrackCanT5Host = *alpaka::getPtrNative(nTrackCanT5Host_buf);
+        trackCandidatesInCPU = new SDL::trackCandidatesBuffer<alpaka::DevCpu>(N_MAX_NONPIXEL_TRACK_CANDIDATES + N_MAX_PIXEL_TRACK_CANDIDATES, devHost, queue);
         trackCandidatesInCPU->setData(*trackCandidatesInCPU);
 
         *alpaka::getPtrNative(trackCandidatesInCPU->nTrackCandidates_buf) = nTrackCanHost;
-        alpaka::memcpy(queue, trackCandidatesInCPU->hitIndices_buf, trackCandidatesBuffers->hitIndices_buf, 14 * nTrackCanHost);
-        alpaka::memcpy(queue, trackCandidatesInCPU->pixelSeedIndex_buf, trackCandidatesBuffers->pixelSeedIndex_buf, nTrackCanHost);
-        alpaka::memcpy(queue, trackCandidatesInCPU->logicalLayers_buf, trackCandidatesBuffers->logicalLayers_buf, 7 * nTrackCanHost);
-        alpaka::memcpy(queue, trackCandidatesInCPU->directObjectIndices_buf, trackCandidatesBuffers->directObjectIndices_buf, nTrackCanHost);
-        alpaka::memcpy(queue, trackCandidatesInCPU->objectIndices_buf, trackCandidatesBuffers->objectIndices_buf, 2 * nTrackCanHost);
-        alpaka::memcpy(queue, trackCandidatesInCPU->trackCandidateType_buf, trackCandidatesBuffers->trackCandidateType_buf, nTrackCanHost);
+        *alpaka::getPtrNative(trackCandidatesInCPU->nTrackCandidatespT5_buf) = nTrackCanpT5Host;
+        *alpaka::getPtrNative(trackCandidatesInCPU->nTrackCandidatespT3_buf) = nTrackCanpT3Host;
+        *alpaka::getPtrNative(trackCandidatesInCPU->nTrackCandidatespLS_buf) = nTrackCanpLSHost;
+        *alpaka::getPtrNative(trackCandidatesInCPU->nTrackCandidatesT5_buf) = nTrackCanT5Host;
+        int nTrackCanToCopy = nTrackCanHost;
+        int nPixTrackCanNotSaved = nTrackCanpT5Host + nTrackCanpT3Host + nTrackCanpLSHost - N_MAX_PIXEL_TRACK_CANDIDATES;
+        int nTrackCanNotSaved = nTrackCanT5Host - N_MAX_NONPIXEL_TRACK_CANDIDATES;
+        nTrackCanToCopy -= (nPixTrackCanNotSaved > 0)*nPixTrackCanNotSaved; // Take into account pT5, pT3, and pLS candidates that were not saved
+        nTrackCanToCopy -= (nTrackCanNotSaved > 0)*nTrackCanNotSaved; // Take into account T5 candidates that were not saved
```

Comment:
So all these math gymnastics, as well as the other kind of convoluted logic that you mentioned in some of the other comments is due to the choice to keep increasing the counter variables (e.g. `nTrackCandidatespT3`) even if the objects are not actually saved. This is what is done in some other parts of the code. It would simplify the logic a lot if we simply stopped saving things and increasing counters once the maximum numbers are reached. But I think it would probably be nice to keep track of the total number of objects even if they don't end up being saved, so maybe a better alternative would be to have additional variables like `nTotalTrackCandidatespT3` that count the total numbers even if they are not saved.